### DISCUSSION TITLE
Ability to disable/bypass caching

### DISCUSSION THREAD

**Discussion by kroymann** [SixLabors/ImageSharp.Web#193] [Ideas]
In older versions of the ImageSharp.Web library, it was possible to effectively disable the caching logic by supplying a dummy IImageCache implementation like this:
```cs
public class NullImageCache : IImageCache
{
    public IDictionary<string, string> Settings => new Dictionary<string, string>();
    public Task<IImageCacheResolver> GetAsync(string key) => Task.FromResult<IImageCacheResolver>(null);
    public Task SetAsync(string key, Stream stream, ImageCacheMetadata metadata) => Task.CompletedTask;
}
```

We leverage this in our application as we already have a custom caching system applied earlier in the request processing pipeline and thus don't need ImageSharp to add any additional caching.  As of version 1.0.3, this technique is no longer possible because the ImageSharpMiddleware was changed to always operate by writing into the cache and then immediately reading back out of the cache again (in commit https://github.com/SixLabors/ImageSharp.Web/commit/2632fc042fc925b5961f3c08d06cbbe0089852f2).

I'd like to add a `DisableCaching` option to the `ImageSharpMiddlewareOptions` that gives consumers an officially supported way of opting out of the caching logic.  Does this seem like a reasonable feature to add, and if so, would anyone object to me putting together a PR?

**Comment by JimBobSquarePants** [comment]
I wasn't aware we'd created that regression. Sorry!

Rather than adding an option I'd like to see if we can have another crack at the caching to allow a `NullImageCache`. It's fair better architecturally that way. If you feel like trying that I'd be very grateful!

**Reply by kroymann** [reply]
I initially tried to find a good way to do just that but given the way the code was changed in the commit I referenced, it's pretty awkward to achieve.  `ProcessRequestAsync()` currently assumes that after writing an image to the cache, it can then follow up with a call to get that same image _from_ the cache and receive valid data.  The response is then being generated by retrieving from the cache the image that was just written into the cache, and the output stream that originally stored the processed image is being disposed of before ever reaching the point where the response is being generated.  I'm not sure I can find a good way to allow the `NullImageCache` without substantially changing the way `ProcessRequestAsync()` works.

A quick read of the ImageSharpMiddleware.cs reveals that a substantial portion of that code is dedicated to the caching logic.  I suspect that having a `DisableCaching` option (or maybe `UseCache` that defaults to true?), and then using that option to branch into a simpler, parallel code path could be the cleanest/safest change to make.

**Reply by kroymann** [reply]
But I'm happy to fiddle around with it and try to come up with a solution either way, as this is currently blocking me from upgrading to the latest version :smile: (I've also been closely following Anton's memory work in the core ImageSharp library and am eager to get my hands on v2.0.0)

**Reply by JimBobSquarePants** [reply]
> a substantial portion of that code is dedicated to the caching logic

I'd say too much. It's all my fault, I should have architected it more loosely.

**Reply by JimBobSquarePants** [reply]
@kroymann I had crack at this by removing the LRU stuff (_since it was holding onto things outside of the cache timing rules and added tons of complexity_) and using a new `PassThroughCacheResolver` to send the response and keeping hold of the output stream the resolver copied from until the write worker result was removed. but I can't seem to crack it. 

The sample site appears to work well but there are several unit tests failing with 
>The status code cannot be set, the response has already started.

That suggests to me something is not awaiting properly but I'm not great with async past the basics. 

The code is here. [js/null-cache](https://github.com/SixLabors/ImageSharp.Web/tree/3e715c07348ecf86b2bb3c8814ff4d0089bb60d3) if you wanna take a look at fixing it.

@deanmarcussen I will need your input here most likely since you figured out how to get this thing to behave in the first place.

**Reply by deanmarcussen** [reply]
@JimBobSquarePants @kroymann 

I seem to recall one of the reasons I had to make the change was the LRU (and possibly disposal of streams was fixed in there at the same time).

Something about when it was in the LRU, it wasn't disposing correctly, and had some memory leaks.

But there were a few fixes around this time, so I can't be 100% sure.

Looking at your branch @JimBobSquarePants where you have removed the awaits from `IsNewOrUpdatedAsync` I think I would try keeping them. That method/arrangement was designed to pause (await) until those tasks were actually complete, so moving the await out of it, and just returning the task, might be changing the behaviour from what we want.

**Reply by kroymann** [reply]
> Looking at your branch @JimBobSquarePants where you have removed the awaits from `IsNewOrUpdatedAsync` I think I would try keeping them. That method/arrangement was designed to pause (await) until those tasks were actually complete, so moving the await out of it, and just returning the task, might be changing the behaviour from what we want.

I don't believe this is it.  The caller of that method is already awaiting the returned Task, so the removal of async/await from `IsNewOrUpdatedAsync()` shouldn't effect the overall execution (although it will make exception callstacks a little bit less clear as it creates a gap in the causality chain - that reason alone is why I've switched to using async/await virtually everywhere, even in simple methods that just thunk through into another async method, but that's a whole difference conversation).

**Reply by kroymann** [reply]
@JimBobSquarePants I see a couple of problems with what you're attempting in your branch.  The biggest is that you are sending the response twice (which is almost assuredly what is causing the error about setting the status code after having already sent the response).  You're doing it once inside the `WriteWorkers.GetOrAddAsync()` callback code on line 371, and then again at the end of the function on line 383.  Another issue is that the current code can result in the disposal of the output stream before other concurrent requests have finished consuming it.  `WriteWorkers` is going to store `ImageWorkerResults` that contain `PassthroughImageCacheResolvers` which in turn store a reference to the output stream.  At the end of the initial request you remove the result from the `WriteWorkers` and dispose of the output stream, but there could easily be other concurrent requests that were waiting for the initial request to complete that will now be trying to copy from the output stream.

**Reply by kroymann** [reply]
If you're open to overhauling the whole ReadWorkers/WriteWorkers pattern, I can take a crack at this.  Let me know if this is an accurate pseudo-code description of the desired logic:
```
if (alreadyCached)
{
    return cachedValue;
}
else
{
    return UseFancyLockToPreventParallelProcessingOfTheSameImage( () =>
    {
        var image = ProcessImage(); 
        StoreResultInCache(image);
        return image;
    };
}
```

NOTE: I laid it out like this because it means that if you don't want to use caching, you could just call `ProcessImage()` directly and skip all the rest of that logic, which kind of fits with what I was suggesting at the start of this discussion.

**Reply by kroymann** [reply]
Alright I put together a branch that re-works the `ImageSharpMiddleware` so that it can return the output stream directly without having to re-read what was just written to the cache.  As part of this, I did away with the ReaderWorkers/WriterWorkers pattern entirely and replaced it with the latest async locking code from my product's codebase.  NOTE: I left the LRU stuff intact and I tried to minimize unnecessary refactoring in order to make the diff as easy to follow as possible.  I also broke the work down into a sequence of step-by-step commits that should help when reviewing it.  Take a look at https://github.com/kroymann/ImageSharp.Web/tree/enable_null_cache and let me know what you think.  I can also make this into a PR if it's easier to comment on the changes that way.

_EDIT: Of course now that I look at the diff in GitHub, I realize that it doesn't ignore whitepace changes (like any proper code diffing tool should allow!), so the changes in indentation make this diff look far more complicated than it actually is :cry:_

**Reply by deanmarcussen** [reply]
Did you bench it?

Putting a lock back in would do this (which was one of the reasons it was removed)
```
if (alreadyCached)
{
    return cachedValue;
}
else
{
// When you lock here, concurrent requests will all have to reprocess the image again, until they have completed
    return UseFancyLockToPreventParallelProcessingOfTheSameImage( () =>
    {
        var image = ProcessImage(); 
        StoreResultInCache(image);
        return image;
    };
}
```

Before I can really offer any sensible comment here, can you explain a little more about why you want to prevent caching?

If you are doing it earlier in the pipeline, can you not plugin your other cache to the ImageSharp cache?

I don't know if this will be relevant, but we have found the Image Sharp metadata that is cached and used for browser caching very important to keep consistent with what is stored in the Image Sharp cache, or browsers get a invalid cache response when checking if the image has changed.

**Reply by kroymann** [reply]
I'll explain more about why I want to be able to bypass ImageSharp's caching in second, but first I want to point out that the async locking that I am trying to re-introduce here is actually quite similar to what the code is already doing.  It relies on `ConcurrentDictionaries` and `Tasks` rather than actual locks, just like the `WriterWorkers` code.  I haven't benchmarked it within ImageSharp but am happy to do so if you can point me at the preferred benchmarking workflow.  Also, for what it's worth, these locking primitives have been in heavy use at the core of Rec Room's servers handling a high volume of concurrent traffic for a few years now (obviously that's not a replacement for benchmarking within the context of ImageSharp, but still, it's a decent vote of confidence).

Regarding your question about concurrent requests having to reprocess the image, I actually addressed that in one of my commits (https://github.com/kroymann/ImageSharp.Web/commit/f804401fa913aacf0d2b5bcfedc00cb6e19bcbc0).  When taking the exclusive writer lock, if the lock wasn't immediately available (which suggests someone else is concurrently processing that image right now), then after obtaining the lock the code re-checks the cache before proceeding to process the image.

Now on the question of why we want to bypass ImageSharp's caching, the TLDR is that we've already got our own layers of caching earlier in the pipeline that provide additional functionality and optimizations.  For example, our cache layer stores custom cryptographic content signatures as additional metadata on the entries and those signatures are included in the response via HTTP headers (we use these signatures to prevent hackers from employing HTTP cache injection attacks as a means to get external imagery such as pornography into Rec Room).

**Reply by JimBobSquarePants** [reply]
I'm curious to see how this runs. The tool used for benchmarking is Crank (though I'll admit I've used it once and usually rely on @deanmarcussen to do it) 

https://docs.microsoft.com/en-us/events/dotnetconf-2021/benchmarking-aspnet-applications-with-net-crank

**Reply by deanmarcussen** [reply]
It might be easier if you created a pr as I haven’t had time to look at the branch at all.
I don’t mind seeing how it performs, but super busy right now. Can take a look early next week.

**Comment by marklagendijk** [comment]
It might be a good idea to add an example of the NullImageCache to the documentation.
Maybe it would even be nice to add it to the library.

For anyone looking for it, if you have nullability enabled in your project the NullImageCache would be:
```c#
public class NullImageCache : IImageCache
{
    public Task<IImageCacheResolver?> GetAsync(string key) => Task.FromResult<IImageCacheResolver?>(null);
    public Task SetAsync(string key, Stream stream, ImageCacheMetadata metadata) => Task.CompletedTask;
}
```
Without nullability:
```c#
public class NullImageCache : IImageCache
{
    public Task<IImageCacheResolver> GetAsync(string key) => Task.FromResult<IImageCacheResolver>(null);
    public Task SetAsync(string key, Stream stream, ImageCacheMetadata metadata) => Task.CompletedTask;
}
```
